File: sdk/go/context.go
package nextloggers

import (
	"context"
	"strings"
)

type logContextKey struct{}

// LogContext is immutable-by-convention request/goroutine context copied into
// each event at creation time. It is safe to capture and re-enter in detached work.
type LogContext struct {
	Fields       map[string]any
	LoggedInUser map[string]any
	Users        []map[string]any
	TraceID      string
	TraceIDs     []string
	RoutineID    string
	Tags         []string
	Context      []any
	Meta         []any
}

func cloneLogContext(source LogContext) LogContext {
	users := make([]map[string]any, 0, len(source.Users))
	for _, user := range source.Users {
		users = append(users, cloneMap(user))
	}
	return LogContext{
		Fields:       cloneMap(source.Fields),
		LoggedInUser: cloneMap(source.LoggedInUser),
		Users:        users,
		TraceID:      source.TraceID,
		TraceIDs:     append([]string(nil), source.TraceIDs...),
		RoutineID:    source.RoutineID,
		Tags:         append([]string(nil), source.Tags...),
		Context:      append([]any(nil), source.Context...),
		Meta:         append([]any(nil), source.Meta...),
	}
}

// MergeLogContext overlays patch onto base without mutating either value.
func MergeLogContext(base, patch LogContext) LogContext {
	merged := cloneLogContext(base)
	for key, value := range patch.Fields {
		merged.Fields[key] = value
	}
	for key, value := range patch.LoggedInUser {
		merged.LoggedInUser[key] = value
	}
	for _, user := range patch.Users {
		merged.Users = append(merged.Users, cloneMap(user))
	}
	if value := strings.TrimSpace(patch.TraceID); value != "" {
		merged.TraceID = value
		merged.TraceIDs = appendUnique(merged.TraceIDs, value)
	}
	for _, traceID := range patch.TraceIDs {
		if value := strings.TrimSpace(traceID); value != "" {
			merged.TraceIDs = appendUnique(merged.TraceIDs, value)
		}
	}
	if patch.RoutineID != "" {
		merged.RoutineID = patch.RoutineID
	}
	for _, tag := range patch.Tags {
		if value := strings.TrimSpace(tag); value != "" {
			merged.Tags = appendUnique(merged.Tags, value)
		}
	}
	merged.Context = append(merged.Context, patch.Context...)
	merged.Meta = append(merged.Meta, patch.Meta...)
	return merged
}

// WithLogContext returns a child context containing a defensive snapshot.
func WithLogContext(parent context.Context, value LogContext) context.Context {
	if parent == nil {
		parent = context.Background()
	}
	if current, ok := LogContextFrom(parent); ok {
		value = MergeLogContext(current, value)
	}
	return context.WithValue(parent, logContextKey{}, cloneLogContext(value))
}

// BackgroundLogContext is shorthand for WithLogContext(context.Background(), value).
func BackgroundLogContext(value LogContext) context.Context {
	return WithLogContext(context.Background(), value)
}

// LogContextFrom returns a defensive copy of the current request context.
func LogContextFrom(ctx context.Context) (LogContext, bool) {
	if ctx == nil {
		return LogContext{}, false
	}
	value, ok := ctx.Value(logContextKey{}).(LogContext)
	if !ok {
		return LogContext{}, false
	}
	return cloneLogContext(value), true
}

// CaptureLogContext snapshots the ambient value for a queue or new goroutine.
func CaptureLogContext(ctx context.Context) LogContext {
	value, _ := LogContextFrom(ctx)
	return value
}

// WithCapturedLogContext re-enters a previously captured value under parent.
func WithCapturedLogContext(parent context.Context, captured LogContext) context.Context {
	return WithLogContext(parent, captured)
}

// ApplyLogContext copies context data into an Event. Later context mutation
// cannot change an already-created record.
func ApplyLogContext(ctx context.Context, event *Event) *Event {
	if event == nil {
		return nil
	}
	value, ok := LogContextFrom(ctx)
	if !ok {
		return event
	}
	for key, field := range value.Fields {
		event.Fields[key] = field
	}
	for key, field := range value.LoggedInUser {
		event.LoggedInUser[key] = field
	}
	for _, user := range value.Users {
		event.Users = append(event.Users, cloneMap(user))
	}
	if value.TraceID != "" {
		event.TraceID = value.TraceID
		event.TraceIDs = appendUnique(event.TraceIDs, value.TraceID)
	}
	for _, traceID := range value.TraceIDs {
		event.TraceIDs = appendUnique(event.TraceIDs, traceID)
	}
	event.RoutineID = value.RoutineID
	for _, tag := range value.Tags {
		event.Tags = appendUnique(event.Tags, tag)
	}
	event.Context = append(event.Context, value.Context...)
	event.Meta = append(event.Meta, value.Meta...)
	return event
}

func (logger *Logger) TraceContext(ctx context.Context, values ...any) *Event {
	return ApplyLogContext(ctx, logger.Trace(values...))
}
func (logger *Logger) DebugContext(ctx context.Context, values ...any) *Event {
	return ApplyLogContext(ctx, logger.Debug(values...))
}
func (logger *Logger) InfoContext(ctx context.Context, values ...any) *Event {
	return ApplyLogContext(ctx, logger.Info(values...))
}
func (logger *Logger) WarnContext(ctx context.Context, values ...any) *Event {
	return ApplyLogContext(ctx, logger.Warn(values...))
}
func (logger *Logger) ErrorContext(ctx context.Context, values ...any) *Event {
	return ApplyLogContext(ctx, logger.Error(values...))
}
func (logger *Logger) FatalContext(ctx context.Context, values ...any) *Event {
	return ApplyLogContext(ctx, logger.Fatal(values...))
}

File: sdk/rust/src/context.rs
//! Runtime-independent scoped logging context.
//!
//! `ContextualFuture` installs the captured frame for each poll, which gives
//! task-local semantics on Tokio, async-std, smol, and custom executors without
//! tying the logger to a global runtime or OpenTelemetry context manager.

use crate::{Event, JsonObject, LogLevel, Logger, Value};
use serde::{Deserialize, Serialize};
use std::cell::RefCell;
use std::future::Future;
use std::pin::Pin;
use std::task::{Context as TaskContext, Poll};

thread_local! {
    static LOG_CONTEXT_STACK: RefCell<Vec<LogContext>> = RefCell::new(Vec::new());
}

#[derive(Clone, Debug, Default, Deserialize, PartialEq, Serialize)]
pub struct LogContext {
    #[serde(default, skip_serializing_if = "JsonObject::is_empty")]
    pub fields: JsonObject,
    #[serde(
        rename = "loggedInUser",
        default,
        skip_serializing_if = "JsonObject::is_empty"
    )]
    pub logged_in_user: JsonObject,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub users: Vec<JsonObject>,
    #[serde(rename = "traceId", skip_serializing_if = "Option::is_none")]
    pub trace_id: Option<String>,
    #[serde(rename = "traceIds", default, skip_serializing_if = "Vec::is_empty")]
    pub trace_ids: Vec<String>,
    #[serde(rename = "routineId", skip_serializing_if = "Option::is_none")]
    pub routine_id: Option<String>,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub tags: Vec<String>,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub context: Vec<Value>,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub meta: Vec<Value>,
}

fn push_unique(values: &mut Vec<String>, value: String) {
    if !value.is_empty() && !values.contains(&value) {
        values.push(value);
    }
}

impl LogContext {
    pub fn merge(mut self, patch: LogContext) -> Self {
        self.fields.extend(patch.fields);
        self.logged_in_user.extend(patch.logged_in_user);
        self.users.extend(patch.users);
        if let Some(trace_id) = patch.trace_id {
            let trace_id = trace_id.trim().to_string();
            if !trace_id.is_empty() {
                self.trace_id = Some(trace_id.clone());
                push_unique(&mut self.trace_ids, trace_id);
            }
        }
        for trace_id in patch.trace_ids {
            push_unique(&mut self.trace_ids, trace_id.trim().to_string());
        }
        if patch.routine_id.is_some() {
            self.routine_id = patch.routine_id;
        }
        for tag in patch.tags {
            push_unique(&mut self.tags, tag.trim().to_string());
        }
        self.context.extend(patch.context);
        self.meta.extend(patch.meta);
        self
    }

    pub fn apply(self, mut event: Event) -> Event {
        event = event.add_fields(self.fields);
        if !self.logged_in_user.is_empty() {
            event = event.add_logged_in_user_info(self.logged_in_user);
        }
        for user in self.users {
            event = event.add_user_info(user);
        }
        if let Some(trace_id) = self.trace_id {
            event = event.add_trace(trace_id, true);
        }
        for trace_id in self.trace_ids {
            event = event.add_trace(trace_id, false);
        }
        if let Some(routine_id) = self.routine_id {
            event = event.add_routine_id(routine_id);
        }
        event = event.add_tags(self.tags);
        for value in self.context {
            event = event.add_context(value);
        }
        for value in self.meta {
            event = event.add_meta(value);
        }
        event
    }
}

struct ContextGuard;

impl Drop for ContextGuard {
    fn drop(&mut self) {
        LOG_CONTEXT_STACK.with(|stack| {
            stack.borrow_mut().pop();
        });
    }
}

/// Runs a synchronous callback inside a nested context frame.
pub fn with_log_context<T>(context: LogContext, callback: impl FnOnce() -> T) -> T {
    LOG_CONTEXT_STACK.with(|stack| stack.borrow_mut().push(context));
    let _guard = ContextGuard;
    callback()
}

pub fn current_log_context() -> Option<LogContext> {
    LOG_CONTEXT_STACK.with(|stack| stack.borrow().last().cloned())
}

/// Captures the current frame for a queue, callback, or spawned task.
pub fn capture_log_context() -> Option<LogContext> {
    current_log_context()
}

pub fn with_captured_log_context<T>(
    captured: Option<LogContext>,
    callback: impl FnOnce() -> T,
) -> T {
    match captured {
        Some(context) => with_log_context(context, callback),
        None => callback(),
    }
}

/// A future that installs one context frame for every poll.
#[must_use = "futures do nothing unless polled or awaited"]
pub struct ContextualFuture<F> {
    context: LogContext,
    future: F,
}

impl<F> ContextualFuture<F> {
    pub fn new(context: LogContext, future: F) -> Self {
        Self { context, future }
    }
}

impl<F: Future> Future for ContextualFuture<F> {
    type Output = F::Output;

    fn poll(self: Pin<&mut Self>, cx: &mut TaskContext<'_>) -> Poll<Self::Output> {
        // SAFETY: once ContextualFuture is pinned, `future` is never moved. We
        // only create a pinned projection for the duration of this poll.
        let this = unsafe { self.get_unchecked_mut() };
        let context = this.context.clone();
        with_log_context(context, || {
            // SAFETY: explained above; the field stays at the same address.
            unsafe { Pin::new_unchecked(&mut this.future) }.poll(cx)
        })
    }
}

pub fn contextualize_future<F: Future>(
    context: LogContext,
    future: F,
) -> ContextualFuture<F> {
    ContextualFuture::new(context, future)
}

pub trait LoggerContextExt {
    fn event_context(&self, level: LogLevel, values: Vec<Value>) -> Event;
    fn trace_context(&self, values: Vec<Value>) -> Event;
    fn debug_context(&self, values: Vec<Value>) -> Event;
    fn info_context(&self, values: Vec<Value>) -> Event;
    fn warn_context(&self, values: Vec<Value>) -> Event;
    fn error_context(&self, values: Vec<Value>) -> Event;
    fn fatal_context(&self, values: Vec<Value>) -> Event;
}

impl LoggerContextExt for Logger {
    fn event_context(&self, level: LogLevel, values: Vec<Value>) -> Event {
        let event = match level {
            LogLevel::Trace => self.trace(values),
            LogLevel::Debug => self.debug(values),
            LogLevel::Info => self.info(values),
            LogLevel::Warn => self.warn(values),
            LogLevel::Error => self.error(values),
            LogLevel::Fatal => self.fatal(values),
        };
        match current_log_context() {
            Some(context) => context.apply(event),
            None => event,
        }
    }

    fn trace_context(&self, values: Vec<Value>) -> Event {
        self.event_context(LogLevel::Trace, values)
    }
    fn debug_context(&self, values: Vec<Value>) -> Event {
        self.event_context(LogLevel::Debug, values)
    }
    fn info_context(&self, values: Vec<Value>) -> Event {
        self.event_context(LogLevel::Info, values)
    }
    fn warn_context(&self, values: Vec<Value>) -> Event {
        self.event_context(LogLevel::Warn, values)
    }
    fn error_context(&self, values: Vec<Value>) -> Event {
        self.event_context(LogLevel::Error, values)
    }
    fn fatal_context(&self, values: Vec<Value>) -> Event {
        self.event_context(LogLevel::Fatal, values)
    }
}
